### PR TITLE
[AI-9418] Close Angie with result via onClose

### PR DESCRIPTION
**Pull request by netanelavr**
## Problem
MCP tools need a way to close Angie and pass a structured result back to the host.

## Solution
1. **`callbacks.onClose(params?)`** — single close callback. Dismiss passes `{}`; close-with-result passes the payload.
2. **`closeAngie(params?, instanceId?)`** — public helper for host MCP tools. Resolves the instance, runs its layout closer, then calls `onClose`.
3. **Host bridge message** `angie/close` (`CLOSE_MESSAGE_TYPE`) — future iframe → host path (needs Angie forwarder).
4. **Demo** in `demo/load-sidebar-v2-css-mcp-app/` — Finish and close tool + `onClose` params panel.

## How it works

```mermaid
sequenceDiagram
    participant MCP as Host MCP tool
    participant Helper as closeAngie
    participant Shell as Layout shell closer
    participant Host as callbacks.onClose

    MCP->>Helper: closeAngie(params, instanceId?)
    Helper->>Shell: closer(params)
    Shell->>Shell: hide UI
    Shell->>Host: onClose(params)
```

Host usage:

```typescript
import { closeAngie } from '@elementor/angie-sdk';

await sdk.loadSidebarV2({
  host: { appId: 'my-app', instanceId: 'my-help-widget' },
  callbacks: {
    onClose: (params = {}) => {
      console.log('Angie closed:', params);
    },
  },
});

// From a host MCP tool:
closeAngie({ reason: 'user-finished' });
// Multi-instance:
closeAngie({ reason: 'user-finished' }, 'my-help-widget');
```

## Test plan
- [x] Unit tests for `closeAngie` / instance targeting / callback throw safety
- [x] Host bridge routes `angie/close` to `closeAngieForInstance`
- [x] Sidebar + floating chat shells notify `onClose`
- [ ] Demo: Finish and close shows params in host UI

```bash
npm run build && npm run demo
# Visit http://localhost:8080/demo/load-sidebar-v2-css-mcp-app/
```

Jira: https://elementor.atlassian.net/browse/AI-9418

## 1. Problem & Context
AI-9418: Enables the host to close Angie programmatically and receive a result object via the `onClose` callback, supporting "close-with-result" workflows from MCP tools.

## 2. What Changed (Where)

| File | Change |
| :--- | :--- |
| `src/load-sidebar-v2/close.ts` | New module managing instance closers and the `closeAngie` public API. |
| `src/index.ts` & `config.ts` | Exported `closeAngie` and updated `onClose` signature to accept `params`. |
| `src/load-sidebar-v2/shell.ts` | Implemented param buffering to sync `closeAngie` results with `onToggle` events. |
| `src/load-sidebar-v2/chat-toggle/chat-shell.ts` | Integrated `notifyClose` and registered instance closer for floating chats. |
| `src/load-sidebar-v2/host-api-bridge.ts` | Added `CLOSE_MESSAGE_TYPE` handler to trigger instance closure via bridge. |
| `demo/...` | Added "Finish and close" tool and UI to demonstrate the flow. |

## 3. How It Works
- **Entry**: Host calls `closeAngie(params, instanceId)` or sends `angie/close` message.
- **Execution**: The `close` module resolves the instance and executes its registered `InstanceCloser`.
- **Sync**: For sidebars, params are buffered in `pendingOnCloseParams` and flushed when `onToggle(false)` triggers, ensuring UI state and callbacks align.

## 4. Risks
- **Callback Exceptions**: `notifyClose` wraps the user callback in a try-catch to prevent host-side crashes from aborting the internal closure logic.

_Generated by LinearB AI and added by gitStream._
<sub>AI-generated content may contain inaccuracies. Please verify before using.
💡 **Tip:** You can customize your AI Description using **Guidelines** [Learn how](https://docs.gitstream.cm/automation-actions/#describe-changes)</sub>

### CODE REVIEW

**Comment by cloud-devpos** [comment]
## Architecture Review Summary
✅ No architecture violations found against applicable catalog principles.

---
*🤖 Architect Review Agent*

**Comment by cloud-devpos** [comment]
## Review Summary
The close-with-params flow is well-structured with solid unit tests and a working demo, but `resolveConfig` never forwards `onCloseWithParams`, so user callbacks registered via `loadSidebarV2({ callbacks: { onCloseWithParams } })` are silently dropped at boot.

### Issues (2)
🏕️ Code Quality: onCloseWithParams dropped during config resolution
🧪 Test Coverage: Missing resolveConfig test for onCloseWithParams

---
*🤖 Code Style Review Agent*

**Comment by cloud-devpos** [comment]
## PR Approver Evaluation

Adds onCloseWithParams callback and MCP App integration for closing Angie with params, with full test coverage and a working demo.

### Reasons to auto-merge
- All CI checks pass: tests on Node 16/18/20, CodeQL, Wiz scanners, and gitstream all green
- Bounded, well-scoped feature: new callback type, message handler, and demo — no unrelated changes bundled in
- New module (close-with-params.ts) is small (33 lines) with clear single responsibility
- Full test coverage: close-with-params.test.ts covers 5 cases including no-instance, no-callback, multi-instance, and default-params edge cases
- host-api-bridge.test.ts extended with 2 integration tests for the new message type
- Follows existing patterns: message type constant exported, registered in boot-sidebar.ts alongside other bridge setup, callback typed in CallbacksConfig
- No auth, secrets, infrastructure, or data-integrity concerns
- Demo files are self-contained in a new demo directory and don't affect production bundle
- Public API exports (closeAngieWithParams, CLOSE_WITH_PARAMS_MESSAGE_TYPE) follow the established export pattern in src/index.ts
- Documentation in README.md is updated with usage example and message ownership table entry

### Reasons not to auto-merge
- closeAngieWithParams always targets the first instance (getFirstInstance()), which may be surprising behavior in multi-instance setups — though this is documented implicitly by the test and mirrors typical single-widget usage
- The demo's close-angie tool posts to a hardcoded ANGIE_ORIGIN ('https://angie.elementor.com') with window.parent.postMessage, which is intentional for the demo context but worth noting
- resetCloseWithParamsForTests is exported from the production module (not a test helper file) — minor code hygiene concern, consistent with existing resetInstancesForTests pattern in the codebase

### Verdict
**Would auto-merge: Yes** — this PR looks low-risk and merge-ready.

---
*🤖 PR Approver Agent*

**Comment by cloud-devpos** [comment]
## PR Approver Evaluation

This PR adds an `onCloseWithParams` callback and supporting `closeAngieWithParams` feature to the load-sidebar-v2 module, plus a new demo app showcasing the flow — all CI checks pass and the scope is well-bounded.

### Reasons to auto-merge
- All CI checks pass: unit tests on Node 16/18/20, CodeQL (JS/TS + Actions), Wiz Data/IaC/SAST/Secret/Vulnerability scanners, and gitStream — pre-check verdict is PASS
- Bounded, cohesive scope: new `close-with-params.ts` module, a new message type constant in `host-api-bridge.ts`, config type extension, and a corresponding demo — no unrelated changes bundled
- New functionality is well-tested: `close-with-params.test.ts` covers callback invocation and sidebar close; `host-api-bridge.test.ts` covers the message dispatch path
- Pattern alignment: follows existing callback conventions (`onClose`, `onToggle`), uses the same `callbacksMap` pattern, registers via `resolveConfig` / `bootSidebar` in line with existing wiring
- Message origin is validated by the existing `host-api-bridge` handler (only messages from the configured `iframeOrigin` are processed), limiting the attack surface of the new `angie/close-with-params` message type
- Test reset helpers (`resetCloseWithParamsForTests`) and spy patterns match project conventions, keeping test isolation clean
- Demo files are additive only and do not affect the production bundle

### Reasons not to auto-merge
- `closeAngieWithParams` uses `getFirstInstance()` — in multi-instance scenarios this silently targets only the first registered widget; no warning or documentation about this limitation
- The `close-angie` MCP tool in the demo posts directly to `ANGIE_ORIGIN` (hardcoded string) via `window.parent.postMessage`, bypassing the SDK abstraction — minor coupling concern in demo code, not production code
- One CI check (`dispatch` in 'PR CI Reviewers' workflow) is still IN_PROGRESS at evaluation time; it is categorised as 'pending' but not a blocker per pre-check verdict

### Verdict
**Would auto-merge: Yes** — this PR looks low-risk and merge-ready.

---
*🤖 PR Approver Agent*

**Comment by cloud-devpos** [comment]
## PR Approver Evaluation

Adds `onCloseWithParams` callback and `closeAngieWithParams` utility so MCP Apps can close Angie and pass structured params back to the host; bounded, well-tested feature addition.

### Reasons to auto-merge
- All CI checks pass (CodeQL, Wiz scanners, tests on Node 16/18/20, gitstream) with no failures
- Scope is tightly bounded: one new message type, one new module (`close-with-params.ts`), a config type extension, and a demo — no unrelated changes
- Unit tests cover both the `closeAngieWithParams` function and the `host-api-bridge` message dispatch path
- Follows existing patterns: new message type constant exported alongside the others, callback wired through `resolveConfig` and `bootSidebar` exactly as `onClose`/`onToggle` are
- No security, auth, secrets, or credential handling involved; message origin is not trusted blindly — the bridge already validates `event.origin === iframeOrigin` upstream
- Demo files are self-contained and clearly labelled; shims (ajv, ajv-formats) are demo-only no-ops that don't affect production code
- README updated accurately to document the new callback and message type
- No infrastructure, deployment, or schema changes

### Reasons not to auto-merge
- `closeAngieWithParams` uses `getFirstInstance()` which means multi-instance setups only notify the first registered instance — acceptable for now but worth a follow-up if multi-widget support is needed
- The `close-angie` MCP tool in the demo calls `window.parent.postMessage` with `ANGIE_ORIGIN` hardcoded; if the origin constant ever diverges from the real iframe origin the message will be silently dropped, but this is demo code only
- No material risks identified in production code paths

### Verdict
**Would auto-merge: Yes** — this PR looks low-risk and merge-ready.

---
*🤖 PR Approver Agent*

**Comment by cloud-devpos** [comment]
## PR Approver Evaluation

Adds a new `closeAngieWithParams` public API and `onCloseWithParams` callback for host MCP tools to close Angie and pass structured params back to the host, with a demo and tests.

### Reasons to auto-merge
- All CI checks passed (CodeQL, tests on Node 16/18/20, Wiz scanners, gitstream)
- Bounded, cohesive scope: new feature adds one callback, one public function, one message type, and wires them together consistently
- New module (close-with-params.ts) has dedicated unit tests covering happy path, multi-instance routing, and callback-throws-but-still-closes edge case
- host-api-bridge.test.ts extended to cover the new CLOSE_WITH_PARAMS_MESSAGE_TYPE message handler
- Follows existing patterns: message type constant exported, config type extended, resolve-config passthrough, boot-sidebar registration call — identical to how other callbacks are wired
- Error in callback is swallowed in a try/catch/finally so the close action is never blocked by host code
- No auth, secrets, credentials, or infrastructure changes
- Demo files are clearly scoped to the demo directory and do not affect production bundle
- CLOSE_WITH_PARAMS_MESSAGE_TYPE exported for consumers who need to reference it externally, consistent with other exported message types

### Reasons not to auto-merge
- closeAngieWithParams (public API) falls back to getFirstInstance(), which silently no-ops in a multi-instance setup if called without knowing which instance to target — documented limitation but could surprise consumers
- The preview-app.html MCP App uses a hardcoded PROTOCOL_VERSION ('2026-01-26') inline rather than importing a shared constant; could drift if the protocol version changes
- Demo shims (ajv-shim.js, ajv-formats-shim.js) always return valid/no-errors regardless of input — acceptable for a demo but worth noting they mask real validation in that context

### Verdict
**Would auto-merge: Yes** — this PR looks low-risk and merge-ready.

---
*🤖 PR Approver Agent*

**Comment by cloud-devpos** [comment]
## PR Approver Evaluation

Adds `closeAngieWithParams` public API for host MCP tools to close Angie and pass structured params back to the host via a new `onCloseWithParams` callback; all CI passes and the change is well-tested.

### Reasons to auto-merge
- All CI checks pass: tests on Node 16/18/20, CodeQL (JS/TS + actions), Wiz scanners (data, IaC, SAST, secret, vulnerability), videoproof, and gitStream.
- Bounded, single-purpose feature: introduces `closeAngieWithParams` + `onCloseWithParams` callback with no unrelated changes bundled in.
- New module (`close-with-params.ts`) is covered by dedicated unit tests including happy path, multi-instance routing, and callback-throws-but-still-closes edge cases.
- Integration test in `host-api-bridge.test.ts` verifies the `angie/close-with-params` message route calls the correct function.
- Callback errors are swallowed in a `try/catch/finally` so a throwing callback can never prevent Angie from closing — correct defensive pattern.
- Public API surface is minimal: exports `closeAngieWithParams` and `CLOSE_WITH_PARAMS_MESSAGE_TYPE`; internal helpers stay unexported from the package root.
- Follows existing patterns: mirrors `onClose` wiring in `resolve-config.ts`, `boot-sidebar.ts`, and `config.ts`; message-type constant naming is consistent with existing bridge constants.
- Documentation (`README.md`) is updated with usage examples and the new callback entry in the config table.
- Demo files are purely additive and illustrative; no production behaviour depends on them.

### Reasons not to auto-merge
- `closeAngieWithParams` silently no-ops when no instance is registered (`getFirstInstance()` returns null) — callers get no error signal; acceptable for a first-pass public API but worth noting.
- Multi-instance `closeAngieWithParams` (public, non-instance variant) always targets the *first* registered instance; could silently target the wrong instance in multi-widget setups, though this is documented as a known limitation.
- The postMessage path (`angie/close-with-params` from nested MCP App) is wired up in the bridge but the README notes the Angie forwarder isn't implemented yet — partial infrastructure is shipped, but it is guarded (only Angie-origin messages reach the bridge) so no premature exposure.
- No material security or data-integrity risks identified.

### Verdict
**Would auto-merge: Yes** — this PR looks low-risk and merge-ready.

---
*🤖 PR Approver Agent*

**Comment by cloud-devpos** [comment]
## PR Approver Evaluation

Adds a `closeAngieWithParams` public API that lets host MCP tools close Angie and pass structured params back to the host callback, with tests and a working demo.

### Reasons to auto-merge
- All CI checks passed (CodeQL, tests on Node 16/18/20, Wiz scanners, videoproof) — only a non-blocking 'dispatch' job is in-progress.
- Bounded, cohesive scope: one new exported function, one new message type, one new callback config field, and corresponding tests + demo — no unrelated changes bundled in.
- New `close-with-params.ts` module is well-structured: uses a Map to support multi-instance, swallows callback errors in a finally block to guarantee the sidebar is always closed, and logs warnings when no instance is found.
- Tests cover the main paths: single-instance, multi-instance targeting, callback-throws-but-still-closes, and the host-api-bridge message dispatch — consistent with the project's existing test patterns.
- New `CLOSE_WITH_PARAMS_MESSAGE_TYPE` constant is exported so consumers can reference the string without hard-coding it.
- Demo app changes are purely additive (new HTML panel, new MCP tool registration, new callback) and do not alter existing demo behaviour.
- README documentation is thorough and correctly describes both the direct host-MCP-tool path and the future nested MCP-App path.
- No auth, secrets, OAuth, permissions, infrastructure, or data-integrity concerns in the changed code.
- No CI or security scanner regressions.

### Reasons not to auto-merge
- `closeAngieWithParams` falls back to `getFirstInstance()` when no `instanceId` is supplied — this is a silent best-effort heuristic that could target the wrong instance in a multi-widget page if the caller forgets to pass the ID; a future improvement could warn more loudly.
- The bridge handler (`CLOSE_WITH_PARAMS_MESSAGE_TYPE` case) does not send a success/error reply to the iframe, unlike most other cases; this is acceptable for a fire-and-forget close but could leave the sender unaware of failure.
- No material risks identified beyond the minor caveats above.

### Verdict
**Would auto-merge: Yes** — this PR looks low-risk and merge-ready.

---
*🤖 PR Approver Agent*

**Comment by cloud-devpos** [comment]
## PR Approver Evaluation

Adds a `closeAngieWithParams` public API that allows host MCP tools to close the Angie sidebar and pass structured params back to the host via a callback, with tests and demo wiring.

### Reasons to auto-merge
- All CI checks passed (tests on Node 16/18/20, CodeQL, Wiz scanners, videoproof)
- Bounded, well-scoped feature: adds one new public function, one new message type, one new callback type, and wires them together without touching unrelated code
- New file `close-with-params.ts` has dedicated unit tests covering single-instance, multi-instance, and callback-throws scenarios
- `host-api-bridge.ts` change is covered by a new test in `host-api-bridge.test.ts`
- Error handling is in place: callback errors are caught/logged and the sidebar is still closed via the `finally` block
- Follows existing patterns: message-type constant export, `initHostApiBridge` handler switch-case, `resolve-config` passthrough, and instance-registry helpers match the surrounding code style
- No auth, secrets, permissions, or credential handling involved
- No infrastructure, CI workflow, Helm, or database changes
- Demo and README documentation updated to match the new feature

### Reasons not to auto-merge
- `window.toggleAngieSidebar` is called unconditionally after the callback — if the sidebar was already closed or the toggle function behaves differently per layout (floating chat vs sidebar), this could produce unexpected UI state, but this mirrors the existing `onClose` pattern so is low risk
- The `CLOSE_WITH_PARAMS_MESSAGE_TYPE` message handler in `host-api-bridge.ts` does not validate or sanitize `event.data.params` before passing it to the callback — arbitrary objects from the iframe origin are forwarded as-is; origin validation is already done upstream so this is acceptable but worth noting
- No material security risks identified beyond the above minor caveats

### Verdict
**Would auto-merge: Yes** — this PR looks low-risk and merge-ready.

---
*🤖 PR Approver Agent*

**Comment by netanelavr** [comment]
## Local demo proof — `closeAngieWithParams` (AI-9418)

Live local run of `demo/load-sidebar-v2-css-mcp-app` on the box (`http://127.0.0.1:8080`), Angie signed in.

### 1. Angie open (signed in)
Angie open signed in

### 3. After close — params on host, Angie closed
After close with params

Received params:

```json
{
  "reason": "user-finished",
  "orderId": "123",
  "timestamp": "2026-09-10T12:59:05.937Z",
  "source": "demo-proof-host-helper"
}
```

Proofs are comment attachments only (not committed under `docs/`).

**Comment by cloud-devpos** [comment]
## PR Approver Evaluation

Adds a `closeAngieWithParams` public API and `onCloseWithParams` callback for host MCP tools to close Angie with structured params, with tests and a demo update.

### Reasons to auto-merge
- All CI checks pass: tests on Node 16/18/20, CodeQL, Wiz scanners, and videoproof are green
- Bounded, well-scoped feature: new `closeAngieWithParams` function, a new `onCloseWithParams` callback type, and the message bridge handler — no unrelated changes
- New file `close-with-params.ts` has 4 unit tests covering: single-instance callback invocation, multi-instance targeting, `closeAngieWithParamsForInstance` directly, and error resilience (callback throws but close still happens)
- Host API bridge test covers the message-dispatch path (`CLOSE_WITH_PARAMS_MESSAGE_TYPE`)
- Error handling is present: callback errors are caught and logged, `window.toggleAngieSidebar` is called in a `finally` block so close always fires
- Public API is exported cleanly through `src/index.ts` with no breaking changes to existing exports
- Changes follow existing patterns: message type constant, bridge `case` handler, callback registration map — all consistent with how `onClose`/`getExternalHeaders` are handled
- Demo code is isolated to `demo/` directory with no production risk
- Documentation in README.md is updated to match the new API

### Reasons not to auto-merge
- One CI check (`PR CI Reviewers` / `dispatch`) is still `IN_PROGRESS` at pre-check time — not a blocker if it passes, but worth confirming it completed green
- `closeAngieWithParams` with no `instanceId` falls back to `getFirstInstance()`, which is a silent best-effort behavior; callers with multiple widgets could silently target the wrong instance if they forget to pass `instanceId`
- The note in the README acknowledges the nested MCP App → host postMessage path is not yet implemented ('requires an Angie forwarder (follow-up)'), so the feature is partially documented as future work

### Verdict
**Would auto-merge: Yes** — this PR looks low-risk and merge-ready.

---
*🤖 PR Approver Agent*

File: src/load-sidebar-v2/resolve-config.ts [hunk 1]
```diff
@@ -64,6 +64,7 @@ export const resolveConfig = ( options: LoadSidebarV2Options, env: Env ): Resolv
 		},
 		callbacks: {
 			onClose: callbacks.onClose,
+			onCloseWithParams: callbacks.onCloseWithParams,
```
Comment (matipojo) [inline src/load-sidebar-v2/resolve-config.ts:67]: Why not use onClose and make the params empty by default

Comment (netanelavr) [reply src/load-sidebar-v2/resolve-config.ts:67]: I kept `onCloseWithParams` separate for intent: `onClose` is a regular dismiss (X/toggle), while this flow closes with a result. Merging them means hosts need to distinguish an empty params object from an actual result.

Same UI, different meaning so separate callback felt cleaner. Happy to switch to overloading `onClose` if we want a smaller API.

**Comment by cloud-devpos** [comment]
## PR Approver Evaluation

Adds closeAngieWithParams public API for host MCP tools to close Angie and pass params back via onClose callback, with tests covering single/multi-instance targeting, error resilience, and sidebar/chat layouts.

### Reasons to auto-merge
- All CI checks pass: tests on Node 16/18/20, CodeQL, Wiz scanners (SAST, secrets, IaC, data, vulnerability), and videoproof all green
- Bounded, coherent scope: one new feature (closeAngieWithParams) with its own module, consistent naming, and no unrelated changes bundled in
- Well-tested: new close-with-params.test.ts (5 unit tests), shell.test.ts extended with 4 new cases (params, resilience, onToggle interaction), chat-shell.test.ts extended with multi-instance targeting test, and host-api-bridge.test.ts covers the postMessage path
- Follows existing patterns: registerInstanceCloser mirrors how the codebase already manages per-instance state; notifyClose wraps onClose with try/catch consistent with host bridge error handling
- Error resilience: onClose throwing does not abort the close; notifyClose catches and logs, window.toggleAngieSidebar failure falls through to direct notify via finally block
- Public API export is additive only (new closeAngieWithParams and CLOSE_WITH_PARAMS_MESSAGE_TYPE exports); onClose type change is backward-compatible (params is optional)
- No security, auth, credentials, infrastructure, or data-integrity concerns in the changed code
- Demo files updated to illustrate the new feature in a contained, illustrative way

### Reasons not to auto-merge
- The 'PR CI Reviewers / dispatch' check is still IN_PROGRESS at pre-check time — if it gates merge it may need to pass first, though it is not blocking CI
- The sidebar closer stashes params in a closure variable (closeParams) and relies on synchronous execution of toggleAngieSidebar to consume it before onToggle fires; an async sidebar toggle implementation could lose the stash, but no such case exists in the current codebase

### Verdict
**Would auto-merge: Yes** — this PR looks low-risk and merge-ready.

---
*🤖 PR Approver Agent*

**Comment by cloud-devpos** [comment]
## PR Approver Evaluation

PR adds a `closeAngie(params, instanceId?)` public API and `callbacks.onClose` params passing, with full test coverage across sidebar and floating chat layouts.

### Reasons to auto-merge
- All CI checks pass: tests on Node 16/18/20, CodeQL, Wiz security scanners, and videoproof all green.
- Bounded, single-purpose scope: adds one new public function (`closeAngie`), one constant (`CLOSE_MESSAGE_TYPE`), and one internal module (`close.ts`) — no cross-cutting rewrites.
- New module (`close.ts`) is well-structured: clear separation between `registerInstanceCloser`, `closeAngieForInstance`, and `closeAngie`; error isolation via `notifyClose` prevents a throwing `onClose` callback from aborting the close.
- Comprehensive test coverage: `close.test.ts` (unit), `shell.test.ts` (sidebar integration), `chat-shell.test.ts` (floating chat integration), and `host-api-bridge.test.ts` (message bridge) all cover the new paths including edge cases (no instance found, callback throws, multi-instance targeting).
- Follows existing patterns: `resetCloseForTests` mirrors the test-reset pattern already used by `resetInstancesForTests` and `resetHostMessageRouterForTests`.
- The `onClose` signature change (`() => void` → `(params?: Record<string, unknown>) => void`) is backwards-compatible; callers ignoring the argument continue to work.
- No auth, secrets, infra, or data-integrity concerns; purely a UI close + callback plumbing change.
- Demo updates are confined to the existing CSS MCP app demo; no production config altered.
- README updated consistently with the new API and message type.

### Reasons not to auto-merge
- The `closeParams` stash-and-consume pattern in `shell.ts` relies on `window.toggleAngieSidebar` being called synchronously. If any async path between `registerInstanceCloser` and `onToggle` is introduced later, params could be lost silently — minor future-maintenance concern, not a current bug.
- The `CLOSE_MESSAGE_TYPE` (`angie/close`) message path in `host-api-bridge` currently only accepts messages from the Angie iframe origin; the README notes a future nested MCP App → host postMessage path is unimplemented, which is fine but worth tracking.
- No material risks identified beyond the above minor caveats.

### Verdict
**Would auto-merge: Yes** — this PR looks low-risk and merge-ready.

---
*🤖 PR Approver Agent*

**Comment by cloud-devpos** [comment]
## PR Approver Evaluation

Adds `closeAngie` public API and `onClose` params passing for close-with-result, with tests and a demo — clearly bounded, safe to auto-merge.

### Reasons to auto-merge
- All CI checks pass (Node 16/18/20, CodeQL, Wiz scanners, videoproof)
- No security, auth, secrets, or infrastructure concerns — purely UI/messaging logic
- Bounded scope: new `close.ts` module plus wiring into existing shell and bridge files
- New public exports (`closeAngie`, `CLOSE_MESSAGE_TYPE`) follow existing patterns in `src/index.ts`
- Well-tested: new unit tests in `close.test.ts`, `shell.test.ts`, and `chat-shell.test.ts` cover single-instance, multi-instance, targeted close, and no-closer fallback
- Graceful degradation: missing closer or no instance logs a warning rather than throwing
- Host callback errors are caught and logged (`notifyClose` try/catch), preventing a throwing `onClose` from aborting the close
- Type signatures updated consistently across `config.ts`, `chat-shell.ts`, and `init-floating-chat-layout.ts`
- Demo and README updated to reflect the new API

### Reasons not to auto-merge
- Sideways close-params flow in `shell.ts` uses a closure-scoped `closeParams` variable consumed by `onToggle` — works correctly but is slightly subtle; a missed `consumeCloseParams()` path could silently drop params (covered by tests today, but worth noting)
- The `CLOSE_MESSAGE_TYPE` postMessage path currently requires the message to come from the Angie iframe origin; the README notes a follow-up is needed for nested MCP App → host postMessage forwarding — not a blocker but a known gap
- No material risks identified beyond the above minor caveats

### Verdict
**Would auto-merge: Yes** — this PR looks low-risk and merge-ready.

---
*🤖 PR Approver Agent*

**Comment by netanelavr** [comment]
## Local demo proof — close-with-params revalidation (box, 2026-09-15)

Re-ran the close checklist on the box against branch `cursor/close-with-params-34ec` (`http://127.0.0.1:8081`). Unit tests 19/19 already green.

MCP App preview card / **Finish and close** button still blocked by known v2 MCP content-client timeout — A1 used the same host API Finish calls (`closeAngieWithParams`) from the page console.

### A1 — Params on host (console stand-in for Finish)
A1 panel + console

```json
{
  "reason": "user-finished",
  "orderId": "456"
}
```

One `Received params from Angie` log; no `Cannot close: instance has no layout shell`.

### A2 — Close via Angie X — JSON preserved
A2

### A3 — Close via host toggle — JSON preserved
A3

### A4 — Second close with params — refreshed
A4

### B — full-config zero-arg `onClose`
B

### C — floating-chat open/close no throw
C

### D — multi-instance: only `demo-chat-2` closed
D

Proofs are comment attachments only (not committed under `docs/` on the PR branch). Screenshots live on `ci/video-proof-assets` like CI video-proof.

**Comment by cloud-devpos** [comment]
## PR Approver Evaluation

Adds a public `closeAngie(params?, instanceId?)` API with `callbacks.onClose(params)` support, backed by a clean instance-closer registry, with full test coverage across all affected modules.

### Reasons to auto-merge
- All CI checks pass (tests on Node 16/18/20, CodeQL, Wiz scanners, videoproof)
- Bounded, well-defined scope: new `close.ts` module + wiring into existing shell and bridge — no unrelated changes
- New `close.ts` has dedicated unit tests covering single-instance, multi-instance targeting, and missing-closer edge cases
- Existing shell and chat-shell tests updated to assert new `params` argument, not just call counts
- Error isolation: `notifyClose` wraps `onClose` in try/catch so a throwing host callback cannot abort the close flow
- Public API surface is minimal: only `closeAngie` and `CLOSE_MESSAGE_TYPE` exported from `src/index.ts`
- Param defaulting is safe: `params ?? {}` prevents undefined from reaching callers
- Demo code is additive only; no production logic changed in the demo files
- Follows existing patterns: instance registry, `resetXForTests` helpers, child logger usage all match codebase conventions

### Reasons not to auto-merge
- The `pendingOnCloseParams` slot in `shell.ts` is a subtle state-passing mechanism between `registerInstanceCloser` and `onToggle`; correct but non-obvious — a code comment explains the intent, which mitigates but doesn't eliminate the risk
- The `CLOSE_MESSAGE_TYPE` ('angie/close') handler in `host-api-bridge.ts` trusts `event.data.params` from the iframe without schema validation; malformed or missing keys pass through silently (low risk given Angie-origin filtering, but worth noting)
- README documents a 'future' postMessage path from nested MCP Apps that is not yet implemented and has no guard; could create confusion if partially relied upon

### Verdict
**Would auto-merge: Yes** — this PR looks low-risk and merge-ready.

---
*🤖 PR Approver Agent*
